File: MATLAB code/Run_with_python/train_model.py

import sys

from U_Net import gen_unet as gen_unet_py
from U_Net import TrainGenerator, ValidGenerator

from keras.callbacks import EarlyStopping

if __name__ == "__main__" : 
    
    model_path = sys.argv[1]
    src_path = sys.argv[2]
    val_src_path = sys.argv[3]
    input_pattern = sys.argv[4]
    output_pattern = sys.argv[5]
    sample_rate = float(sys.argv[6])
    epoch = int(sys.argv[7])

    model_path = model_path + "/" if model_path[-1] != "/" else model_path

    train_gen = TrainGenerator(
        src_path, input_pattern, output_pattern, 
        sample_rate=sample_rate
    )
    val_gen = ValidGenerator(
        val_src_path, input_pattern, output_pattern, 
        sample_rate=sample_rate
    )
    print("Input shape : ", train_gen.input_shape)

    model = gen_unet_py(input_shape=train_gen.input_shape)
    model.compile(optimizer="adam", loss="mae")

    print("========================= Model summary =========================\n")
    print(model.summary())
    print("\n=================================================================\n")
    
    print("Training model may take times...")

    patience = int(epoch / 10)
    if patience == 0 : patience = 1
    callback = EarlyStopping(verbose=False, restore_best_weights=True, patience=patience)

    print("Start training...", end="")
    model.fit(
        x=train_gen, validation_data=val_gen, verbose=False, 
        epochs=epoch, callbacks=callback
    )
    print("\t--> Done")
    model.save(model_path + "model.h5")
    print("Model was saved in [{}]".format(model_path + "model.h5"))

File: MATLAB code/Run_with_python/train_model.m

% Args
% ------------------------------------------------------------------------------------- %
% 1. model_save_path    :   path to save trained model.                                 %
% 2. src_path           :   path where train data exists.                               %
% 3. val_src_path       :   path where validation data exists.                          %
%                                                                                       %
% 4. input_pattern      :   data pattern to use input training data.                    %
%                           if input_pattern="*.wav", it will use every .wav files as   %
%                           input data.                                                 %
%                                                                                       %
% 5. output_pattern     :   data pattern to use output training data.                   %
%                           if output_pattern="*.wav", it will use every .wav files as  %
%                           input data.                                                 %
%                                                                                       %
% 6. sample_rate        :   sample rate to resolve data.                                %
% 7. epoch              :   epochs to train data.                                       %
% ------------------------------------------------------------------------------------- %

function train_model(model_save_path, src_path, val_src_path, ...
    input_pattern, output_pattern, sample_rate, epoch)
    
    args = model_save_path + " " + src_path + " " + val_src_path + " " ...
        + input_pattern + " " + output_pattern + " " ...
        + sample_rate + " " + epoch;

    args = "train_model.py " + args;

    pyrunfile(args)
end